internal/impl/gcp/input_spanner.go

```diff
@@ -162,26 +183,79 @@ func newSpannerReader(conf spannerInputConfig, mgr *service.Resources) (*spanner
 	return r, nil
 }
 
-//------------------------------------------------------------------------------
+func (r *spannerReader) onReadResult(res *changestreams.ReadResult) error {
+	b, err := json.Marshal(res.ChangeRecords)
+	if err != nil {
+		return fmt.Errorf("failed to marshal read result as JSON: %w", err)
+	}
+
+	msg := service.NewMessage(b)
+	msg.MetaSetMut("spanner_partition_token", res.PartitionToken)
+	msg.MetaSetMut("spanner_project_id", r.conf.ProjectID)
+	msg.MetaSetMut("spanner_instance_id", r.conf.InstanceID)
+	msg.MetaSetMut("spanner_database_id", r.conf.DatabaseID)
+	msg.MetaSetMut("spanner_stream_id", r.conf.StreamID)
+
+	r.resCh <- msg
+
+	return nil
+}
 
 func (r *spannerReader) Connect(ctx context.Context) error {
 	r.log.Infof("Connecting to mock Spanner CDC stream: %s (project: %s, instance: %s, database: %s)",
 		r.conf.StreamID, r.conf.ProjectID, r.conf.InstanceID, r.conf.DatabaseID)
 
+	var err error
+	r.reader, err = changestreams.NewReaderWithConfig(
+		ctx,
+		r.conf.ProjectID,
+		r.conf.InstanceID,
+		r.conf.DatabaseID,
+		r.conf.StreamID,
+		changestreams.Config{
+			StartTimestamp:    r.conf.StartTimestamp,
+			EndTimestamp:      r.conf.EndTimestamp,
+			HeartbeatInterval: r.conf.HeartbeatInterval,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create Spanner change stream reader: %w", err)
+	}
+
+	r.resCh = make(chan *service.Message)
+
+	var bg context.Context
+	bg, r.cancel = context.WithCancel(ctx)
+	go func() {
+		r.connected.Store(true)
+		if err := r.reader.Read(bg, r.onReadResult); err != nil {
+			r.log.Errorf("Error reading from Spanner change stream: %v", err)
+		}
+		r.connected.Store(false)
+	}()
+
+	return nil
+}
+
+func (r *spannerReader) ack(ctx context.Context, err error) error {
 	return nil
 }
 
 func (r *spannerReader) Read(ctx context.Context) (*service.Message, service.AckFunc, error) {
-	if ctx.Err() != nil {
-		return nil, nil, ctx.Err()
+	if !r.connected.Load() {
+		return nil, nil, service.ErrNotConnected
 	}
```

Comment:
Since these are inherently a bit racy, we generally use `shutdown.Signaller` to know when the connection drops and when the background fiber has stopped.